// stringreader/dist/stringreader.js
'use strict';

var babelHelpers = {};

babelHelpers.classCallCheck = function (instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
};

babelHelpers;

var StringReader = function () {
  function StringReader(input) {
    babelHelpers.classCallCheck(this, StringReader);

    this.input = input;
    this.reset();
  }

  StringReader.prototype.reset = function reset() {
    this.currentIndex = 0;
    this.tokens = [];
    this.clearBuffer();
  };

  StringReader.prototype.clearBuffer = function clearBuffer() {
    this.buffer = '';
  };

  // TODO: Can we get a generator here?


  StringReader.prototype.readUntil = function readUntil(character, omitCharacter, notFoundFn) {
    /* eslint-disable no-constant-condition */

    this.clearBuffer();

    var omitChar = omitCharacter == null ? false : omitCharacter;

    var counter = 0;

    if (!omitChar) {
      if (this.input.substr(this.currentIndex, character.length) === character) {
        this.currentIndex += character.length;
        this.tokens.push(character);
        return character;
      }
    }

    // Not liking the contant here :'(
    while (true) {
      if (this.end()) {
        if (notFoundFn) {
          notFoundFn(this.buffer);
        }
        break;
      }

      if (character.length === 1) {
        if (this.input[this.currentIndex] === character) {
          break;
        }
      } else {
        var skip = true;
        for (var i = 0, len = character.length; i < len; i++) {
          var c = character[i];

          if (!(this.input[this.currentIndex + counter] !== c)) {
            continue;
          }
          skip = false;
          break;
        }
        counter++;
        if (skip) {
          break;
        }
      }
      this.buffer += this.input[this.currentIndex];
      this.currentIndex++;
    }

    if (omitChar) {
      this.currentIndex += character.length;
    }

    this.tokens.push(this.buffer);

    return this.buffer;
  };

  StringReader.prototype.readLine = function readLine(notFoundFn) {
    return this.readUntil('\n', true, notFoundFn);
  };

  StringReader.prototype.end = function end() {
    return this.currentIndex >= this.input.length;
  };

  return StringReader;
}();

module.exports = StringReader;
